config/backend-config/backend-config.go

```diff
@@ -388,23 +383,53 @@ func (bc *CommonBackendConfig) WaitForConfig(ctx context.Context) error {
 	return nil
 }
 
-// Setup backend config
-func Setup(configEnvHandler types.ConfigEnvI) {
-	if isMultiWorkspace && misc.IsMultiTenant() {
-		backendConfig = new(MultiTenantWorkspaceConfig)
-		backendConfig.(*MultiTenantWorkspaceConfig).CommonBackendConfig.configEnvHandler = configEnvHandler
-	} else if isMultiWorkspace {
-		backendConfig = new(MultiWorkspaceConfig)
-	} else {
-		backendConfig = new(WorkspaceConfig)
-		backendConfig.(*WorkspaceConfig).CommonBackendConfig.configEnvHandler = configEnvHandler
+func newForDeployment(deploymentType deployment.Type, configEnvHandler types.ConfigEnvI) (BackendConfig, error) {
+	var backendConfig BackendConfig
+
+	switch deploymentType {
+	case deployment.DedicatedType:
+		backendConfig = &SingleWorkspaceConfig{
+			CommonBackendConfig: CommonBackendConfig{
+				configEnvHandler: configEnvHandler,
+			},
+		}
+	case deployment.HostedType:
+		backendConfig = &HostedWorkspacesConfig{}
+	case deployment.MultiTenantType:
+		backendConfig = &MultiTenantWorkspacesConfig{
+			CommonBackendConfig: CommonBackendConfig{
+				configEnvHandler: configEnvHandler,
+			},
+		}
+	// Fallback to dedicated
+	default:
+		return nil, fmt.Errorf("Deployment type %q not supported", deploymentType)
 	}
 
 	backendConfig.SetUp()
```

Comment:
given the way we are currently using SetUp & IsConfigured, it is more appropriate to:

1. Remove `IsConfigured`
2. Change `SetUp`'s signature to return an `error`
3. Return an error if `SetUp` fails, e.g. didn't manage to setup a token or whatever